File: 023 HackerRank.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <algorithm>

using namespace std;

vector<int> get_proper_divisor_sums(int limit) {
    vector<int> divisor_sums(limit + 1, 1);
    divisor_sums[0] = 0;
    for (int i = 2; i <= limit / 2; ++i) {
        for (int j = 2 * i; j <= limit; j += i) {
            divisor_sums[j] += i;
        }
    }
    return divisor_sums;
}

vector<int> get_abundant_numbers(const vector<int>& divisor_sums, int limit) {
    vector<int> abundants;
    for (int i = 1; i <= limit; ++i) {
        if (i < divisor_sums[i]) {
            abundants.push_back(i);
        }
    }
    return abundants;
}

unordered_set<int> get_sum_of_two_abundants(const vector<int>& abundants, int limit) {
    unordered_set<int> sum_of_2_abundants;
    for (size_t i = 0; i < abundants.size(); ++i) {
        for (size_t j = i; j < abundants.size(); ++j) {
            int sum = abundants[i] + abundants[j];
            if (sum <= limit) {
                sum_of_2_abundants.insert(sum);
            } else {
                break;
            }
        }
    }
    return sum_of_2_abundants;
}

int main() {
    int n;
    cin >> n;
    vector<int> number_list(n);
    int limit = 0;
    
    for (int i = 0; i < n; ++i) {
        cin >> number_list[i];
        limit = max(limit, number_list[i]);
    }
    
    vector<int> divisor_sums = get_proper_divisor_sums(limit);
    vector<int> abundants = get_abundant_numbers(divisor_sums, limit);
    unordered_set<int> sum_of_2_abundants = get_sum_of_two_abundants(abundants, limit);
    
    for (int num : number_list) {
        if (sum_of_2_abundants.find(num) != sum_of_2_abundants.end()) {
            cout << "YES\n";
        } else {
            cout << "NO\n";
        }
    }
    
    return 0;
}

File: 023 HackerRank.py
from itertools import combinations_with_replacement

### inputs
number_list = []
for _ in range(int(input())):
    number_list.append(int(input()))
limit = max(number_list)

### get the sum of proper divisors of each number below the limit separately
l = [1]*(limit+1)
l[0] = 0
for i in range(2,limit//2+1):
    for j in range(2*i,limit+1,i):
        l[j] = l[j]+i

### create a list of abundant numbers
abundants = [ i for i in range(1,limit) if i < l[i] ]

### finding all the positive integers which can be written as the sum of two abundant numbers
sum_of_2_abundants = set(i+j for i,j in combinations_with_replacement(abundants,2) if i+j<=limit)


for number in number_list:
    if number in sum_of_2_abundants:
        print('YES')
    else:
        print('NO')

